sdk/opendp/whitenoise/synthesizers/pytorch/nn/dpctgan.py

```diff
@@ -0,0 +1,331 @@
+
+import ctgan
+from ctgan.transformer import DataTransformer
+from ctgan.conditional import ConditionalGenerator 
+from ctgan.models import Discriminator, Generator
+from ctgan.sampler import Sampler 
+
+from ctgan import CTGANSynthesizer
+
+import torchdp
+from torchdp import autograd_grad_sample
+from torchdp import PrivacyEngine, utils
+
+
+
+def __init__(self, input_dim, dis_dims, loss, pack):
+        super(Discriminator, self).__init__()
+        torch.cuda.manual_seed(0)
+        torch.manual_seed(0)
+        
+        dim = input_dim * pack
+      #  print ('now dim is {}'.format(dim))
+        self.pack = pack
+        self.packdim = dim
+        seq = []
+        for item in list(dis_dims):
+            seq += [Linear(dim, item), LeakyReLU(0.2), Dropout(0.5)]
+            dim = item
+
+        seq += [Linear(dim, 1)]
+        if loss == 'cross_entropy':
+            print ("this should happen")
+            seq += [Sigmoid()]
+        self.seq = Sequential(*seq)
+        
+Discriminator.__init__ = __init__
+
+
+
+
+
+
+
+class DPCTGAN(SDGYMBaseSynthesizer, CTGANSynthesizer):
+    """Differential Private Conditional Table GAN Synthesizer
+    This code adds Differential Privacy to CTGANSynthesizer from https://github.com/sdv-dev/CTGAN 
+    """
+
+    def __init__(self,
+                 embedding_dim=128,
+                 gen_dim=(256, 256),
+                 dis_dim=(256, 256),
+                 l2scale=1e-6,
+                 batch_size=500,
+                 epochs=300,
+                 pack=1,
+                 log_frequency=True,
+                 disabled_dp=False,
+                 target_delta=None,
+                 sigma = 5,
+                 max_per_sample_grad_norm=1.0,
+                 budget = 3,
+                 verbose=True,
+                 loss = 'cross_entropy'
+
+                 ):
+
+        # CTGAN model specific parameters
+        self.embedding_dim = embedding_dim
+        self.gen_dim = gen_dim
+        self.dis_dim = dis_dim
+
+        self.l2scale = l2scale
+        self.batch_size = batch_size
+        self.epochs = epochs
+        self.pack=pack
+        self.log_frequency = log_frequency
+        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        # torchdp parameters
+        self.sigma = sigma
+        self.disabled_dp = disabled_dp
+        self.target_delta = target_delta
+        self.max_per_sample_grad_norm = max_per_sample_grad_norm
+        self.budget = budget
+        self.epsilon_list = []
+        self.alpha_list = []
+        self.loss_d_list = []
+        self.loss_g_list = []
+        self.verbose=verbose
+        self.loss=loss
+
+
+    def train(self, train_data, categorical_columns=tuple(), ordinal_columns=tuple()):
+
+        
+        self.transformer = DataTransformer()
+        self.transformer.fit(train_data, categorical_columns)
+        train_data = self.transformer.transform(train_data)
```

Comment:
You should be able to remove this and call PytorchDPSynthesizer(DataTransformer(), DPCTGAN()), where the output of DataTransformer is the numpy array that gets fed into the GAN network.

Reply:
> You should be able to remove this and call PytorchDPSynthesizer(DataTransformer(), DPCTGAN()), where the output of DataTransformer is the numpy array that gets fed into the GAN network.

similar reason as before regarding why do preprocessing data/transforming data in train step